jobs/periodic/delete-unreachable-achievements/courses.ts

```diff
@@ -0,0 +1,78 @@
+import { getLogger } from '../../../common/logger/logger';
+import { EventRelationType, createRelation, parseRelation } from '../../../common/achievement/relation';
+import { prisma } from '../../../common/prisma';
+import tracer from '../../../common/logger/tracing';
+
+const logger = getLogger('CourseAchievementCleanup');
+
+const ParticipationGroupTemplate = 'pupil_course_participation';
+
+export const deleteUnreachableCourseAchievements = tracer.wrap('delete', _deleteUnreachableCourseAchievements);
+async function _deleteUnreachableCourseAchievements(dryRun: boolean = true) {
+    const till = new Date();
+    till.setDate(till.getDate() - 1);
+
+    const unachievedAchievementes = await prisma.user_achievement.findMany({
+        where: {
+            template: {
+                group: ParticipationGroupTemplate,
+            },
+            achievedAt: null,
+        },
+    });
+
+    const relationToUserIds: { [key: string]: string[] } = unachievedAchievementes.reduce((acc, curr) => {
+        if (!acc[curr.relation]) {
+            acc[curr.relation] = [];
+        }
+        acc[curr.relation].push(curr.userId);
+        return acc;
+    }, {});
+
+    logger.info('found possibly unachievable achievements', {
+        courses: relationToUserIds,
+        search_till: till,
+    });
+
+    for (const [relation, userIds] of Object.entries(relationToUserIds)) {
+        const [relationType, subcourseId] = parseRelation(relation);
+        if (relationType !== EventRelationType.Subcourse) {
+            logger.warn('unexpected relation type', {
+                relation,
+            });
+            continue;
+        }
+
+        const futureLectureCnt = await prisma.lecture.count({
```

Comment:
Use https://github.com/corona-school/backend/blob/33c9681b9deae57856ff37f0f3ce8c3cb5217c7b/common/courses/states.ts#L65 ? Then we have a single place to decide (also there is the plan to have "infinite courses" which never end)

Reply:
Here you could also easily add `|| subcourseCancelled(...)` :)

Reply:
That's a great idea. I didn't know about these functions. I'll add them.